network/p2p/tracer/gossipSubScoreTracer.go

```diff
@@ -0,0 +1,253 @@
+		updateInterval: updateInterval,
+		collector:      collector,
+		snapshotUpdate: make(chan struct{}, 1),
+		snapshot:       make(map[peer.ID]*p2p.PeerScoreSnapshot),
+		idProvider:     provider,
+	}
+
+	g.Component = component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			g.logLoop(ctx)
+		}).
+		Build()
+
+	return g
+}
+
+// UpdatePeerScoreSnapshots updates the tracer's snapshot of the peer scores.
+func (g *GossipSubScoreTracer) UpdatePeerScoreSnapshots(snapshot map[peer.ID]*p2p.PeerScoreSnapshot) {
+	g.snapshotLock.Lock()
+	// critical section
+	g.snapshot = snapshot
+	g.snapshotLock.Unlock()
+
+	g.requestSnapshotUpdate() // notify the log loop that there is a new snapshot.
+}
+
+// UpdateInterval returns the interval at which the tracer expects to receive updates from the gossipsub router.
+func (g *GossipSubScoreTracer) UpdateInterval() time.Duration {
+	return g.updateInterval
+}
+
+// GetScore returns the overall score for the given peer.
+func (g *GossipSubScoreTracer) GetScore(peerID peer.ID) (float64, bool) {
+	g.snapshotLock.RLock()
+	defer g.snapshotLock.RUnlock()
+
+	if snapshot, ok := g.snapshot[peerID]; ok {
+		return snapshot.Score, true
+	}
+
+	return 0, false
+}
+
+// GetAppScore returns the application score for the given peer.
+func (g *GossipSubScoreTracer) GetAppScore(peerID peer.ID) (float64, bool) {
+	g.snapshotLock.RLock()
+	defer g.snapshotLock.RUnlock()
+
+	if snapshot, ok := g.snapshot[peerID]; ok {
+		return snapshot.AppSpecificScore, true
+	}
+
+	return 0, false
+}
+
+// GetIPColocationFactor returns the IP colocation factor for the given peer.
+func (g *GossipSubScoreTracer) GetIPColocationFactor(peerID peer.ID) (float64, bool) {
+	g.snapshotLock.RLock()
+	defer g.snapshotLock.RUnlock()
+
+	if snapshot, ok := g.snapshot[peerID]; ok {
+		return snapshot.IPColocationFactor, true
+	}
+
+	return 0, false
+}
+
+// GetBehaviourPenalty returns the behaviour penalty for the given peer.
+func (g *GossipSubScoreTracer) GetBehaviourPenalty(peerID peer.ID) (float64, bool) {
+	g.snapshotLock.RLock()
+	defer g.snapshotLock.RUnlock()
+
+	if snapshot, ok := g.snapshot[peerID]; ok {
+		return snapshot.BehaviourPenalty, true
+	}
+
+	return 0, false
+}
+
+// GetTopicScores returns the topic scores for the given peer.
+// The returned map is keyed by topic name.
+func (g *GossipSubScoreTracer) GetTopicScores(peerID peer.ID) (map[string]p2p.TopicScoreSnapshot, bool) {
+	g.snapshotLock.RLock()
+	defer g.snapshotLock.RUnlock()
+
+	topicsSnapshot := make(map[string]p2p.TopicScoreSnapshot)
+	snapshot, ok := g.snapshot[peerID]
+	if !ok {
+		return nil, false
+	}
+
+	// copy the topic scores into a new map
+	for topic, topicSnapshot := range snapshot.Topics {
+		topicsSnapshot[topic] = *topicSnapshot
+	}
+
+	return topicsSnapshot, true
+}
+
+func (g *GossipSubScoreTracer) logLoop(ctx irrecoverable.SignalerContext) {
```

Comment:
call `g.logPeerScores()` once before logLoop exits

Reply:
It's worth noting that calling the logger before the loop exit can create a liveness vulnerability. If the logger has a termination issue and gets stuck, the node may be unable to shut down gracefully because the loop will not exit. This can cause the node to become unresponsive and potentially cause other issues.